core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala

```
@@ -307,6 +309,103 @@ class MultiCloudWorkspaceService(override val ctx: RawlsRequestContext,
       } yield profileModel
     }
 
+  def cloneMultiCloudWorkspaceAsync(
+    wsService: WorkspaceService,
+    sourceWorkspaceName: WorkspaceName,
+    destWorkspaceRequest: WorkspaceRequest
+  ): Future[Workspace] =
+    for {
+      sourceWs <- getV2WorkspaceContextAndPermissions(sourceWorkspaceName, SamWorkspaceActions.read)
+      billingProject <- getBillingProjectContext(RawlsBillingProjectName(destWorkspaceRequest.namespace))
+      _ <- requireCreateWorkspaceAction(billingProject.projectName)
+      billingProfileOpt <- getBillingProfile(billingProject)
+      clone <- (sourceWs.workspaceType, billingProfileOpt) match {
+
+        case (McWorkspace, Some(profile)) if profile.getCloudPlatform == CloudPlatform.AZURE =>
+          traceFutureWithParent("cloneAzureWorkspace", ctx) { s =>
+            cloneAzureWorkspaceAsync(sourceWs, profile, destWorkspaceRequest, s)
+          }
+
+        case (RawlsWorkspace, profileOpt)
+            if profileOpt.isEmpty ||
+              profileOpt.map(_.getCloudPlatform).contains(CloudPlatform.GCP) =>
+          traceFutureWithParent("cloneRawlsWorkspace", ctx) { s =>
+            wsService.cloneWorkspace(sourceWs, billingProject, destWorkspaceRequest, s)
+          }
+
+        case (wsType, profileOpt) =>
+          Future.failed(
+            RawlsExceptionWithErrorReport(
+              ErrorReport(
+                StatusCodes.BadRequest,
+                s"Cloud platform mismatch: Cannot clone $wsType workspace '$sourceWorkspaceName' " +
+                  s"into billing project '${billingProject.projectName}' " +
+                  s"(hosted on ${profileOpt.map(_.getCloudPlatform).getOrElse(CloudPlatform.GCP)})."
```

Nit:The hardcoding of the fallback to GCP is a bit awkward here; I wonder if we should just throw rather than potentially returning misinformation.

---

It's an error message, so I can just make it default to a string. I'll change it in the original location as well.